### PR TITLE
EOS-25961: [IAM Policy] Policy Validations added

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Problem Statement
- Problem statement
- Validate IAM Policies before they are getting created and stored.
[EOS-25710](https://jts.seagate.com/browse/EOS-25710) 
[EOS-25961](https://jts.seagate.com/browse/EOS-25961)
# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by rohitkolapkar** [comment]
@bkirunge7 @shalakadharap 
All the code changes requested in https://github.com/Seagate/cortx-s3server/pull/1595 are addressed in this current PR. 
Please review.

**Comment by rohitkolapkar** [comment]
S3Server-PreMerge-Test 
http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/601/

File: auth/server/src/main/java/com/seagates3/policy/IAMPolicyValidator.java [hunk 1]
```diff
@@ -20,25 +20,197 @@
...
+          for (int count = 0; count < arr.length(); count++) {
+            JSONObject obj = (JSONObject)arr.get(count);
+            response = validateStatementElements(obj);
+            if (response != null) {
+              return response;
+            }
+          }
+        } else if (jsonObject.get(
+                       JsonDocumentFields.STATEMENT) instanceof JSONObject) {
+          LOGGER.debug("Validating policy statement object.");
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/policy/IAMPolicyValidator.java:155]: There are too many debug statements, add only one if required

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java [hunk 2]
```diff
@@ -68,7 +77,7 @@ ServerResponse validateEffect(Effect effectValue) {
       if (Statement.Effect.valueOf(effectValue.name()) == null) {
         response = responseGenerator.malformedPolicy("Invalid effect : " +
                                                      effectValue.name());
-        LOGGER.error("Effect value is invalid in bucket policy");
+        LOGGER.error("Effect value is invalid in policy");
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:80]: print policy name

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java [hunk 3]
```diff
@@ -229,4 +238,170 @@ boolean isPrincipalIdValid(Principal principal) {
...
+  }
+ protected
+  ServerResponse validateStatementSyntax(JSONObject jsonObject) {
+    ServerResponse response = null;
+    if (jsonObject.get(JsonDocumentFields.STATEMENT) instanceof JSONArray) {
+      JSONArray arr = (JSONArray)jsonObject.get(JsonDocumentFields.STATEMENT);
+      if (arr.length() == 0) {
+        response =
+            responseGenerator.malformedPolicy("Syntax errors in policy.");
+        LOGGER.error("Statement array can not be empty");
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:272]: print policy name

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java (hunk 2 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:80]: Policy name is not available here. 
But I have added the debug log with policy name included at the time of calling validation from PolicyController#Create.

File: auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java (hunk 3 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/policy/PolicyValidator.java:272]: Policy name is not available here. 
But I have added the debug log with policy name included at the time of calling validation from PolicyController#Create.

File: auth/server/src/main/java/com/seagates3/policy/IAMPolicyValidator.java (hunk 1 above)
Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/policy/IAMPolicyValidator.java:155]: These are condition based debug statements, only one of them will get added in logs.

**Review by bkirunge7 — APPROVED** [review]

File: st/clitests/policy_files/iam-policy.json [hunk 4]
```diff
@@ -8,7 +7,7 @@
         "s3:PutBucketAcl"
       ],
       "Effect": "Allow",
-      "Resource": "arn:aws:s3:::samplebucket"
+      "Resource": "arn:aws:iam::samplebucket"
```
Comment (shalakadharap) [inline st/clitests/policy_files/iam-policy.json:10]: Wrong change.
PutBucketAcl is S3 service so previous arn is correct.
plz change it original

File: auth/server/src/main/java/com/seagates3/policy/IAMArnParser.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.seagates3.policy;
+
+public
+class IAMArnParser extends ArnParser {
+ public
+  IAMArnParser() {
+
+    this.regEx = "arn:aws:iam::[a-zA-Z0-9~@#$^*\\\\\\\\/_.:-]+";
```
Comment (shalakadharap) [inline auth/server/src/main/java/com/seagates3/policy/IAMArnParser.java:8]: Why you not using ArnParser

Comment (rohitkolapkar) [reply auth/server/src/main/java/com/seagates3/policy/IAMArnParser.java]: fixed

File: st/clitests/policy_files/iam-policy.json (hunk 4 above)
Comment (rohitkolapkar) [reply st/clitests/policy_files/iam-policy.json]: done

**Review by shalakadharap — APPROVED** [review]
Changes looking fine.
Address codacy warnings.

**Pull request merged**